rpc/RPCTransaction.java

```diff
@@ -135,45 +129,157 @@ public void close() {
+        return iterateAsync(request, transformResponse).get();
     }
 
-    public <T> QueryFuture<Stream<T>> iterateAsync(final TransactionProto.Transaction.Iter.Req iterRequest, final Function<TransactionProto.Transaction.Iter.Res, T> responseReader) {
-        try (ThreadTrace trace = traceOnThread("iterateAsync")) {
-            final TransactionProto.Transaction.Req request = TransactionProto.Transaction.Req.newBuilder()
-                    .setIterReq(iterRequest).build();
-            final RPCResponseAccumulator collector = new RPCResponseAccumulator();
-            final RPCIterator<T> iterator = new RPCIterator<>(request, responseReader, requestObserver, collector);
-            collectorQueue.add(collector);
-            return new QueryFuture.Stream<>(iterator);
+    public <T> QueryFuture<Stream<T>> iterateAsync(TransactionProto.Transaction.Req.Builder request, Function<TransactionProto.Transaction.Res, T> transformResponse) {
+        try (ThreadTrace ignored = traceOnThread("iterateAsync")) {
+            final ResponseCollector.Multiple responseCollector = new ResponseCollector.Multiple();
+            final UUID requestId = UUID.randomUUID();
+            request.setId(requestId.toString());
+            request.setLatencyMillis(networkLatencyMillis);
+            collectors.put(requestId, responseCollector);
+            return new QueryFuture.Stream<>(request.build(), requestObserver, responseCollector, transformResponse);
         }
     }
 
-    private synchronized void dispatchRequest(final TransactionProto.Transaction.Req request, final RPCResponseAccumulator collector) {
-        // We must add the response collectors in the exact same order we send the requests
-        // TODO: this doesn't seem like it should be true
-        collectorQueue.add(collector);
-        requestObserver.onNext(request);
+    private StreamObserver<TransactionProto.Transaction.Res> responseObserver() {
+        return new StreamObserver<TransactionProto.Transaction.Res>() {
+            @Override
+            public void onNext(TransactionProto.Transaction.Res res) {
+                final UUID requestId = UUID.fromString(res.getId());
+                final ResponseCollector collector = collectors.get(requestId);
+                if (collector == null) throw new GraknClientException(UNKNOWN_REQUEST_ID.message(requestId));
+                collector.put(new Response.Ok(res));
+                if (collector.isDone()) collectors.remove(requestId);
+            }
+
+            @Override
+            public void onError(Throwable error) {
+                assert error instanceof StatusRuntimeException : "The server only yields these exceptions";
+                // TODO: is it desirable behaviour to kill all the pending requests?
+                // TODO: this isn't nice in any case - the error for other pending requests is misleading
```

Comment:
No I think we should talk about them

Reply:
Raised https://github.com/graknlabs/client-java/issues/180

Reply:
This is now resolved.

The correct behaviour is that both `onError` and `onCompleted` should send an error signal to all consumers which will cause them to throw an exception if they attempt to keep consuming. `onCompleted` can occur as the result of a successful `commit` on the server. This will cause any response collectors open on the client side to cease to be able to receive further responses. So we send them just one more response, which when the user attempts to read it, throws an exception saying the transaction has been closed. Also we set `isOpen` to false, which will prevent the RPCTransaction attempting to close the RPC stream that has already been closed.

Nonetheless the error in `onError` may not be appropriate for these failing consumers - see #180 .